### PR TITLE
add ~GraphToDictMapper~ UsersCollector

### PR DESCRIPTION
**Pull request by matthiasdiener**

### CODE REVIEW

File: pytato/transform.py [hunk 1]
```diff
@@ -988,4 +989,126 @@ def materialize_with_mpms(expr: DictOfNamedArrays) -> DictOfNamedArrays:
 
 # }}}
 
+
+# {{{ graph-to-dict
+
+class GraphToDictMapper(Mapper):
```
Comment (kaushikcfd) [inline pytato/transform.py:995]: Probably derive from CachedMapper?

Comment (kaushikcfd) [inline pytato/transform.py:995]: GraphToDict --> UsersCollector? (This is quite similar to pt.analysis.NUserCollector)

File: pytato/transform.py [hunk 2]
```diff
@@ -988,4 +989,126 @@ def materialize_with_mpms(expr: DictOfNamedArrays) -> DictOfNamedArrays:
...
+    """
+    Maps a graph to a dictionary representation mapping a node to its parents,
```
Comment (kaushikcfd) [inline pytato/transform.py:997]:
```suggestion
    Maps a graph to a dictionary representation mapping a node to its users,
```

File: pytato/transform.py [hunk 3]
```diff
@@ -988,4 +989,126 @@ def materialize_with_mpms(expr: DictOfNamedArrays) -> DictOfNamedArrays:
...
+    i.e. all the nodes using its value.
+    .. attribute:: graph_dict
+    """
+
+    def __init__(self) -> None:
+        """Initialize the GraphToDictMapper."""
+        self.graph_dict: Dict[ArrayOrNames, Set[ArrayOrNames]] = {}
+
+    def map_dict_of_named_arrays(self, expr: DictOfNamedArrays, *args: Any) -> None:
```
Comment (kaushikcfd) [inline pytato/transform.py:1006]: I don't think there is a good need to derive from this mapper, as its quite specific. So drop `*args` from all mapper methods?

**Review by kaushikcfd — COMMENTED** [review]
Thanks!

File: pytato/transform.py (hunk 3 above)
Comment (matthiasdiener) [reply pytato/transform.py]: Done in 1065b07

File: pytato/transform.py (hunk 1 above)
Comment (matthiasdiener) [reply pytato/transform.py]: Done in 1065b07

Comment (matthiasdiener) [reply pytato/transform.py]: I see the similarity to the `NUserCollector`, but as far as I can see "user" isn't really used anywhere in pytato to describe the "children"... Imho, GraphToDict is a bit clearer.

File: pytato/transform.py (hunk 2 above)
Comment (matthiasdiener) [reply pytato/transform.py:997]: (see above regarding "user".)

File: pytato/transform.py (hunk 1 above)
Comment (kaushikcfd) [reply pytato/transform.py]: A problem with using a term like parent, child (or even successor/predecessor) is ambiguous as we haven't defined the convention of the direction of the edges. I think [User](https://llvm.org/doxygen/classllvm_1_1User.html) is a more aligned term wrt the literature.

File: pytato/transform.py [hunk 4]
```diff
@@ -59,6 +59,7 @@
 .. autoclass:: CachedWalkMapper
 .. autoclass:: TopoSortMapper
 .. autoclass:: CachedMapAndCopyMapper
+.. autoclass:: GraphToDictMapper
```
Comment (inducer) [inline pytato/transform.py:62]: Separate section? Add `tag_child_nodes` and `reverse_graph`.

File: pytato/transform.py [hunk 5]
```diff
@@ -988,4 +989,128 @@ def materialize_with_mpms(expr: DictOfNamedArrays) -> DictOfNamedArrays:
...
+        for other_node_key in edges:
+            result.setdefault(other_node_key, set()).add(node_key)
+
+    return result
+
+
+def tag_child_nodes(graph: Dict[Array, Set[Array]], tag: Any,
+        starting_point: Optional[Array] = None,
+        node_to_tags: Optional[Dict[Optional[Array], Set[Array]]] = None) -> None:
+    """Tags nodes reachable from *starting_point*."""
```
Comment (inducer) [inline pytato/transform.py:1105]: Better docstring.

File: pytato/transform.py [hunk 6]
```diff
@@ -988,4 +989,128 @@ def materialize_with_mpms(expr: DictOfNamedArrays) -> DictOfNamedArrays:
...
+                self.graph_dict.setdefault(dim, set()).add(expr)
+                self.rec(dim)
+
+# }}}
+
+
+# {{{ operations on graphs in dict form
+
+def reverse_graph(graph: Dict[Array, Set[Array]]) -> Dict[Array, Set[Array]]:
+    """Reverses a graph."""
```
Comment (inducer) [inline pytato/transform.py:1092]: Better docstring.

File: pytato/transform.py (hunk 1 above)
Comment (inducer) [reply pytato/transform.py]: I don't mind `User` terminology-wise.

File: test/test_pytato.py [hunk 7]
```diff
@@ -304,6 +304,23 @@ def test_toposortmapper():
     assert isinstance(tm.topological_order[6], MatrixProduct)
 
 
+def test_graphtodictmapper():
+    n = pt.make_size_param("n")
+    array = pt.make_placeholder(name="array", shape=n, dtype=np.float64)
+    stack = pt.stack([array, 2*array, array + 6])
+    y = stack @ stack.T
```
Comment (inducer) [inline test/test_pytato.py:311]: Use a few random DAGs instead.

Comment (matthiasdiener) [reply test/test_pytato.py]: 382e9cf

File: pytato/transform.py (hunk 1 above)
Comment (matthiasdiener) [reply pytato/transform.py]: Renamed in 5633775

File: pytato/transform.py (hunk 5 above)
Comment (matthiasdiener) [reply pytato/transform.py]: e165444

File: pytato/transform.py (hunk 6 above)
Comment (matthiasdiener) [reply pytato/transform.py]: e165444

File: pytato/transform.py (hunk 4 above)
Comment (matthiasdiener) [reply pytato/transform.py]: e165444

**Comment by matthiasdiener** [comment]
Thank you for the comments, this is ready for another review @kaushikcfd @inducer .

**Pull request merged**